### PR TITLE
Add Vulkan Backend

### LINKED ISSUE #85: Vulkan Backend
## Background
`yakui` is great! Really cool! Great UI library. But you know what would make it _even cooler_?

image

An extra _10,000 lines of redundant code_!

## TODO
- [x] Research `yakui-app`, `yakui-wgpu` `yakui-winit`
- [x] Create stub renderer
- [x] Build the damn thing

### PR DESCRIPTION
**Pull request by kanerogers**
## Background
Closes #85

### CODE REVIEW

**Comment by kanerogers** [comment]
Alright @LPGhatguy - I think this is ready to go!

File: crates/yakui-vulkan/shaders/main.frag [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#version 450
+#extension GL_ARB_separate_shader_objects : enable
+#extension GL_ARB_shading_language_420pack : enable
```
Comment (Ralith) [inline crates/yakui-vulkan/shaders/main.frag:3]: I don't think either of these do anything here.

**Review by Ralith — CHANGES_REQUESTED** [review]

**Comment by kanerogers** [comment]
Rats, thought I’d killed those. Thanks!

**Review by LPGhatguy — CHANGES_REQUESTED** [review]
I got tied up with some stuff over this last week and didn't finish my review. Here's what I've got for you so far, though!

File: crates/yakui-vulkan/src/vulkan_context.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+///
+/// ## Safety
+/// It is **very** important that you pass the correct handles to this struct, or you will have a terrible time.
+/// Once you create a [`crate::YakuiVulkan`] instance, you **must** use the same handles each time you call a
+/// method on that instance.
+///
+/// See the documentation on each member for specific safety requirements.
+pub struct VulkanContext<'a> {
+    /// A valid Vulkan device that has the `VkPhysicalDeviceDescriptorIndexingFeatures.descriptorBindingPartiallyBound` feature enabled
+    pub device: &'a ash::Device,
```
[the remark this answers is not in the document]
Comment (Ralith) [reply crates/yakui-vulkan/src/vulkan_context.rs:16]: `Device` has most of the function pointer tables, so it's pretty big.

File: .vscode/tasks.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+				"yakui-vulkan"
+			],
+			"problemMatcher": [
+				"$rustc"
+			],
+			"group": "test",
+			"label": "rust: cargo test"
+		}
+	]
+}
```
[the remark this answers is not in the document]
Comment (kanerogers) [reply .vscode/tasks.json:18]: Yes! Good point. Removed.

File: crates/yakui-vulkan/src/vulkan_context.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+/// A wrapper around handles into your Vulkan renderer to call various methods on [`crate::YakuiVulkan`]
+///
+/// ## Safety
+/// It is **very** important that you pass the correct handles to this struct, or you will have a terrible time.
+/// Once you create a [`crate::YakuiVulkan`] instance, you **must** use the same handles each time you call a
+/// method on that instance.
+///
+/// See the documentation on each member for specific safety requirements.
+pub struct VulkanContext<'a> {
+    /// A valid Vulkan device that has the `VkPhysicalDeviceDescriptorIndexingFeatures.descriptorBindingPartiallyBound` feature enabled
```
[the remark this answers is not in the document]
Comment (kanerogers) [reply crates/yakui-vulkan/src/vulkan_context.rs:15]: Yep, absolutely.

File: crates/yakui-vulkan/src/vulkan_texture.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    memory: vk::DeviceMemory,
+    pub sampler: vk::Sampler,
+    pub view: vk::ImageView,
+    pub id: u32,
+}
+
+/// A container for information about a texture.
+/// Populate this struct and pass it to [`crate::YakuiVulkan::add_user_texture()`] to create user managed
+/// textures that you can then use in [`yakui`] code.
+pub struct VulkanTextureCreateInfo<T: AsRef<[u8]>> {
```
[the remark this answers is not in the document]
Comment (kanerogers) [reply crates/yakui-vulkan/src/vulkan_texture.rs:18]: > Bounds on structs express the wrong thing for most people. They are infectious, redundant, sometimes nearsighted, and often confusing.

Yikes. :P Okay, point taken!

File: crates/yakui-vulkan/src/vulkan_context.rs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            vk::SubmitInfo::builder().command_buffers(std::slice::from_ref(&command_buffer));
+        device
+            .queue_submit(self.queue, std::slice::from_ref(&submit_info), fence)
+            .unwrap();
+        device
+            .wait_for_fences(std::slice::from_ref(&fence), true, u64::MAX)
+            .unwrap();
+
+        device.destroy_fence(fence, None);
+        device.free_command_buffers(self.command_pool, std::slice::from_ref(&command_buffer));
```
[the remark this answers is not in the document]
Comment (kanerogers) [reply crates/yakui-vulkan/src/vulkan_context.rs:188]: Yeah, it's pretty neat - something I stole from @Ralith! In my own code I generally alias it to something like `slice_from_ref` but I just wanted to be a bit clearer here.

[the remark this answers is not in the document]
Comment (kanerogers) [reply crates/yakui-vulkan/src/vulkan_context.rs:188]: Yeah, it's pretty neat - something I stole from @Ralith! In my own code I generally alias it to something like `slice_from_ref` but I just wanted to be a bit clearer here.

File: crates/yakui-vulkan/src/vulkan_texture.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        device.destroy_sampler(self.sampler, None);
+        device.destroy_image_view(self.view, None);
+        device.destroy_image(self.image, None);
+        device.free_memory(self.memory, None);
+    }
+}
+
+fn get_format(yakui_format: yakui::paint::TextureFormat) -> vk::Format {
+    match yakui_format {
+        yakui::paint::TextureFormat::Rgba8Srgb => vk::Format::R8G8B8A8_UNORM,
```
[the remark this answers is not in the document]
Comment (kanerogers) [reply crates/yakui-vulkan/src/vulkan_texture.rs:124]: As dicussed earlier, yes, it aboslutely should be! But I did have some strangeness with how images were being decoded with the `image` crate. I'll dig a little deeper into this today.

**Review by LPGhatguy — APPROVED** [review]
After keeping you hanging for forever (sorry!) I'm going to go ahead and merge this as-is. I'll poke at it afterwards and we can always land follow-up changes.

**Pull request merged**